commerce/domain/promotion/cart/calculator/classes/PromotionCalculatorSample.cls

```diff
@@ -0,0 +1,98 @@
+ /**
+  * @description This sample is for the situations where Promotion Calculation needs to be extended or overridden via the
+  * extension point for the Promotion Calculator. You are expected to refer this and write your own implementation.
+  * This class must extend the CartExtension.PromotionsCartCalculator class to be processed.
+  */
+public with sharing class PromotionCalculatorSample extends CartExtension.PromotionsCartCalculator {
+
+    // You MUST change this to be a valid promotion id.
+    public static final String DUMMY_PROMOTION_ID = '0c8xx000000003FAAQ';
+
+    public virtual override void calculate(CartExtension.CartCalculateCalculatorRequest request) {
```

Comment:
@patricia-bagzai-sf This method would always reset adjustments on the whole cart and reapply all of them. This has time complexity O(n) where n is the number of cart items. In most of the cases buyer just modifies one item in the cart at a time. We would prefer to not recalculate items that were not modified in scope of given request. In order to achieve this CartCalculateCalculatorRequest has BuyerActionDetails which describes the specific change on the Cart that you could use for optimization. More information available here: https://salesforce.quip.com/2gSJAlXoWHJE#ZDQABA7HGDz
https://salesforce.quip.com/8X5fAhaNEODJ#temp:C:TFafa0875e04eac4a9cb3f1dee45

Otherwise is there a specific reason why BuyerActionDetails are not used in this public sample?

Reply:
Hello @pplat23 a couple points on the above.

1. We cant look at just the cart items that change when evaluating promotions, please refer to : https://salesforce.quip.com/Ly49AvET3dfB#temp:C:TdBa5cd0e64e87140f181776a1e0 for details. So I dont think using BuyerActionDetails in this example really makes sense.
2. We can probably get rid of resetAllAdjustments for this example if we want to improve performance, but the example's intention is just to demonstrate which fields to set and making it "performant" was not really the intent.
3. Down the road to really improve performance / ease implementation we should probably expose an APEX method to "Merge Adjustments" that does this in a performant way, similar to how we do with our own calculator.

Reply:
@curtwalker1 thank you for getting back to me.

> We cant look at just the cart items that change when evaluating promotions, please refer to : https://salesforce.quip.com/Ly49AvET3dfB#temp:C:TdBa5cd0e64e87140f181776a1e0 for details. So I dont think using BuyerActionDetails in this example really makes sense.

The explanation in quip doc makes sense for complex promotions rules. However, this sample is clearly different - it applies flat 5% discount on every cart item. I don't see why you need to reapply it over and over again even given cart item has not been changed.

> We can probably get rid of resetAllAdjustments for this example if we want to improve performance, but the example's intention is just to demonstrate which fields to set and making it "performant" was not really the intent.

I strongly feel we should have an intention to demonstrate how to write "performant" extension. If we don't demonstrate - customer don't do it. That's a simple as that.

> Down the road to really improve performance / ease implementation we should probably expose an APEX method to "Merge Adjustments" that does this in a performant way, similar to how we do with our own calculator.

That sounds like a great idea! As we envision complex promotions calculations is one of the heaviest operations during the checkout.

With this being said, I'm happy that we have this example available and I would be grateful if you guys have a plan to accommodate the feedback I provided!

Reply:
Hi @pplat23, thank you for sharing the points. We have created a [WI](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE00001kIlxVYAS/view) to address this & will fix it as part of it. Thanks!